### DISCUSSION TITLE
Nested element locator differences

### DISCUSSION THREAD

**Discussion by Auto81** [appium/appium#17967] [Q&A]
appium server 2 beta 44
appium java-client : 8.3.0

android uiAutomator2@2.12.0 testing a Native App

selenium/grid 3
selenium/grid 4.7.2

I'm in the process of updating to selenium/grid4 and have come across the following. Not sure if its appium or selenium, so if it's not appium i will move it to selenium repo.

**Through Grid 3**

> When finding an element (using xpath) within an element, everything behaves as expected

> When finding an element (using ID) within an element, everything behaves as expected

**Through Grid 4 relay to the same Grid 3**

> When finding an element (using xpath) within an element, everything behaves as expected

> When finding an element (using ID) within an element, the wrong element is returned. It will be the first matching element on page source. IE it doesn't seem to be taking the parent element into account.

`13:06:05.351 WARN [CustomLocatorHandler.execute] - Custom conversion to a CSS locator from 'id' will be removed soon. Please switch to a valid W3C WebDriver locator strategy https://www.w3.org/TR/webdriver1/#locator-strategies
`
[Linked to locator strategies here](https://www.w3.org/TR/webdriver1/#locator-strategies)

ID doesn't seem to be a supported w3c locator (which seems weird by itself). I'm guessing the custom conversion that selenium 4 is doing from ID to CSS is ignoring the parent? Kind of like xpath '`.//*`' vs '`//*`'

`[debug] [AndroidUiautomator2Driver@5b55 (4ab6b2d3)] Calling AppiumDriver.findElementFromElement() with args: ["css selector","#com\\.company\\:id\\/customID","00000000-0000-57e9-ffff-ffff0000017f","4ab6b2d3-cded-4b3d-821e-e9a7e50d9f88"]
`

Is this a known thing for appium with grid 4? Or will the suggested fix going forward to be, supply a CSS locator rather than ID locator?

Thanks for time

**Comment by KazuCocoa** [comment]
Appium still needs non W3C locators such as `id` since native apps do not have `css`, which it is mostly for Web. So, the current expected behavior is Grid 4 won't change anything.
https://github.com/SeleniumHQ/selenium/commit/cd3e9aea5261fb9ea371b95abe8ae6113241dfb6 is the relevant change.
Expected behavior is Grid won't change anything about the request from the client, I think.

**Reply by Auto81** [reply]
So it looks like the grid 4 **is** currently changing the ID locator to a CSS locator

ie java-client code is using the locator
`By.id("com.company:id/customID")`

But the grid4/relay is intercepting and converting it to a css selector and logs out the following
`WARN [CustomLocatorHandler.execute] - Custom conversion to a CSS locator from 'id' will be removed soon. Please switch to a valid W3C WebDriver locator strategy https://www.w3.org/TR/webdriver1/#locator-strategies`

Is there a flag that needs set somewhere to disable this attempted conversion?

**Reply by jlipps** [reply]
do you know @diemol ?

**Reply by diemol** [reply]
Grid is doing that because Simon had the idea to help users move to W3C WebDriver by translating locator calls from non-W3C compliant clients to W3C compliant locators. We have revisited that change and the first step was to remove the limitation of not sending non-W3C locators (like accesibilityId and others). However, we do not remove things before warning users, and that is why you see that message in the logs. I believe it has been more than two releases since we added that log message, so we should be fine to remove those translations. @Auto81, if you create an issue in the Selenium repo, I can track it and then remove those translators.

**Reply by Auto81** [reply]
Brilliant, thanks for this.

Created issue over in selenium repo here  - https://github.com/SeleniumHQ/selenium/issues/11470
